usecases/schema/schema_repair.go

```diff
@@ -0,0 +1,276 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package schema
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/weaviate/weaviate/entities/models"
+	"github.com/weaviate/weaviate/entities/schema"
+	"github.com/weaviate/weaviate/usecases/cluster"
+	"github.com/weaviate/weaviate/usecases/sharding"
+)
+
+func (m *Manager) repairSchema(ctx context.Context, remote *State) error {
+	m.logger.WithField("action", "repair_schema").
+		Debug("Attempting to repair schema")
+	before := time.Now()
+
+	local := &m.schemaCache.State
+	if local.ObjectSchema == nil || remote.ObjectSchema == nil {
+		return fmt.Errorf("nil schema found, local: %v, remote: %v",
+			local.ObjectSchema, remote.ObjectSchema)
+	}
+
+	if err := (&clusterSyncRepairs{}).start(ctx, m, local, remote); err != nil {
+		return err
+	}
+
+	m.logger.WithField("action", "repair_schema").
+		Debugf("Schema repair complete, took %s", time.Since(before))
+	return nil
+}
+
+func (m *Manager) repairLocalClasses(ctx context.Context, classes []*models.Class) error {
+	for _, class := range classes {
+		shardState, err := sharding.InitState(class.Class,
+			class.ShardingConfig.(sharding.Config),
+			m.clusterState, class.ReplicationConfig.Factor,
+			schema.MultiTenancyEnabled(class))
+		if err != nil {
+			return fmt.Errorf("init sharding state: %w", err)
+		}
+		if err = m.addClassApplyChanges(ctx, class, shardState); err != nil {
+			return fmt.Errorf("add class %q locally: %w", class.Class, err)
+		}
+	}
+	return nil
+}
+
+func (m *Manager) repairLocalProperties(properties map[string][]*models.Property) error {
+	for class, props := range properties {
+		for _, prop := range props {
+			_, err := m.schemaCache.addProperty(class, prop)
+			if err != nil {
+				return fmt.Errorf(`add prop "%s.%s" locally: %w`, class, prop.Name, err)
+			}
+		}
+	}
+	return nil
+}
+
+func (m *Manager) repairLocalTenants(tenants map[string][]sharding.Physical) {
+	m.schemaCache.LockGuard(func() {
+		for class, tenants := range tenants {
+			ss := m.schemaCache.State.ShardingState[class]
+			if ss.Physical == nil {
+				ss.Physical = make(map[string]sharding.Physical)
+			}
+			for _, tenant := range tenants {
+				ss.Physical[tenant.Name] = tenant
+			}
+		}
+	})
+}
+
+func (m *Manager) repairRemoteClasses(ctx context.Context, classes []*models.Class) error {
+	for _, class := range classes {
```

Comment:
For future improvements, we might want to consider batch repairing classes, properties, and tenants

Reply:
Yes, this would be a big improvement by reducing the number of transactions needed